pkg/healthz/healthz_test.go

```diff
@@ -0,0 +1,359 @@
+							},
+							OnStoppedLeading: func() {
+								log.Info("leaderelection.LeaderCallbacks.OnStoppedLeading")
+							},
+							OnNewLeader: func(identity string) {
+								log.Info("leaderelection.LeaderCallbacks.OnNewLeader", "identity", identity)
+								if tc.in.onNewLeaderHook != nil {
+									tc.in.onNewLeaderHook(&tc.in)
+								}
+								close(leaderElected)
+							},
+						},
+						Lock:     tc.in.resourceLock,
+						WatchDog: s.LeaderHealthzAdaptor,
+					})
+					return nil
+				})
+			}
+
+			log.Info("Starting healthz server go-routine")
+			g.Go(func() error {
+				defer log.Info("Healthz server go-routine finished")
+				return s.Start(gCTX, l)
+			})
+
+			if tc.in.leaderElectionEnabled {
+				log.Info("Waiting for a LeaderElector to know the current leader before polling liveness endpoint")
+				<-leaderElected
+			}
+
+			const (
+				pollingInterval = 500 * time.Millisecond
+				pollingTimeout  = 3 * time.Second
+			)
+			log.Info(
+				"Polling liveness endpoint",
+				"url", livezURL,
+				"interval", pollingInterval,
+				"timeout", pollingTimeout,
+			)
+			var (
+				lastResponseCode int
+				lastResponseBody string
+			)
+			assert.Eventually(t, func() bool {
+				resp, err := http.Get(livezURL)
+				require.NoError(t, err)
+				defer func() {
+					require.NoError(t, resp.Body.Close())
+				}()
+				bodyBytes, err := ioutil.ReadAll(resp.Body)
+				require.NoError(t, err)
+
+				lastResponseCode = resp.StatusCode
+				lastResponseBody = string(bodyBytes)
+
+				log.Info("liveness-probe", "response-code", lastResponseCode, "response-body", lastResponseBody)
+
+				return tc.out.responseCode == lastResponseCode && tc.out.responseBody == lastResponseBody
+			}, pollingTimeout, pollingInterval)
+
+			assert.Equal(t, tc.out.responseBody, lastResponseBody)
+			assert.Equal(t, tc.out.responseCode, lastResponseCode)
+			cancel()
+			require.NoError(t, g.Wait())
+		})
+	}
+}
+
+// fakeResourceLock implements resourcelock.Interface sufficiently to simulate:
+// * successful acquisition of the leader election lock by the local node,
+// * current possession of the leader election lock by a remote node, and
+// * failures in leader election which cause the `LeaderElection.Run` function to exit.
+//
+// The intention is to be able to test the behavior of the
+// LeaderElectionHealthzAdaptor under those circumstances.
+type fakeResourceLock struct {
+	record      *resourcelock.LeaderElectionRecord
+	getError    error
+	updateError error
+}
+
+func (o *fakeResourceLock) Identity() string {
+	return localIdentity
+}
+
+func (o *fakeResourceLock) Describe() string {
+	return lockDescription
+}
+
+func (o *fakeResourceLock) Get(ctx context.Context) (*resourcelock.LeaderElectionRecord, []byte, error) {
+	klog.FromContext(ctx).WithName("fakeResourceLock").Info("Get")
+	if o.getError != nil {
+		return nil, nil, o.getError
+	}
+	if o.record == nil {
+		err := errors.NewNotFound(schema.ParseGroupResource("configmap"), "foo")
```

Comment:
can we change this to a `lease`, I don't think we use configmaps anymore

Reply:
It's a simulated error so not really important. The LeaderElector code only checks for `errors.IsNotFound`. It does not care what resource type is.

https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/leaderelection/leaderelection.go#L326-L332